base/src/Data/Macaw/Memory/ElfLoader.hs

```diff
@@ -1353,54 +1427,94 @@ instance Ord (MemSymbol w) where
+                          , Bounded  (Elf.ElfWordType w)
+                          , Integral (Elf.ElfWordType w)
+                          , Show (ElfWordType w)
+                          , MemWidth w) => a)
+                     -> a
+elfHeaderConstraints elf = reprConstraints (elfAddrWidth (Elf.headerClass (Elf.header elf)))
+
 {-# DEPRECATED memoryForElf' "Use memoryForElf" #-}
 memoryForElf' :: LoadOptions
-              -> Elf w
-              -> Either String ( Memory w
-                               , SectionIndexMap w
-                               , [MemLoadWarning]
-                               )
-memoryForElf' opt e = reprConstraints (elfAddrWidth (elfClass e)) $ do
-  case Elf.elfType e of
-    -- We load object files by section
-    Elf.ET_REL -> do
-      when (isJust (loadOffset opt)) $ do
-        Left $ "Object file sections have multiple offsets, and do not support loading at address."
-      memoryForElfSections e
-    -- Other files are loaded by segment.
-    _ ->  memoryForElfSegments opt e
+              -> Elf.ElfHeaderInfo w
+              -> (Elf.SymtabEntry BS.ByteString (ElfWordType w) -> Bool)
+              -- ^ Filter on symbol table entries
+              -> Either String  ( Memory w
+                                , [MemSymbol w] -- Function symbols
+                                , [MemLoadWarning]
+                                , [SymbolResolutionError]
+                                )
+memoryForElf' opt elf isRelevant = elfHeaderConstraints elf $ do
+  let shdrs = Elf.headerShdrs elf
+  let contents = Elf.headerFileContents elf
+  ((mem, secMap, warnings), dynamicEntries) <-
+    case Elf.headerType (Elf.header elf) of
+      -- We load object files by section
+      Elf.ET_REL -> do
+        when (isJust (loadOffset opt)) $ do
+          Left $ "Object file sections have multiple offsets, and do not support loading at address."
+        r <- memoryForElfSections elf
+        pure (r, V.empty)
+      Elf.ET_EXEC -> do
+        r <- memoryForElfSegments opt elf
+        pure (r, V.empty)
+      -- Dynamic include dynamic information
+      _ -> do
+        r <- memoryForElfSegments opt elf
+
+        -- Get dynamic entries
+        let hdr = Elf.header elf
+        let cl = Elf.headerClass hdr
+        let dta = Elf.headerData hdr
+
+
+
+        let dynamicEntries
+              | [symtab] <- V.toList $ V.filter (\s -> Elf.shdrType s == Elf.SHT_DYNSYM) shdrs
+              , strtabIdx <- Elf.shdrLink symtab
+              , Just strtab <- shdrs V.!? fromIntegral strtabIdx
+              , symtabData <- shdrData contents symtab
+              , strtabData <- shdrData contents strtab
+              , Right entries <-Elf.decodeSymtab cl dta strtabData symtabData =
+                  entries
+              | otherwise = V.empty
+        pure (r, dynamicEntries)
+  let (symErrs, funcSymbols) = resolveElfFuncSymbols mem shdrs secMap isRelevant dynamicEntries elf
```

Comment:
That code is confusing -- I'll add a comment to the code to clarify.
This code is only incorporating the dynamicEntries because resolveElfFuncSymbols knows how to resolve static symbols, but doesn't try to resolve dynamic symbols.  The dynamic symbols are only relevant to the `ET_DYN` elf type because that is the only type that should contain dynamic symbol tables.